### ISSUE TITLE
Allow the addition of Dashboard context to any layout

### ISSUE THREAD

**Issue by argiepiano** [backdrop/backdrop-issues#6182]
## Description of the need

Currently, the Dashboard context is automatically added to layouts when the path is `admin/dashboard`. This allows the layout to access Dashboard-specific blocks.

<img width="477" alt="Screen Shot 2023-07-24 at 2 21 15 PM" src="https://github.com/backdrop/backdrop-issues/assets/9938978/dbdd7496-c821-47de-a174-cbbab180580d">

It'd be great to be able to add this context to layouts with other paths. Currently this context is not listed among the options for other paths. 

<img width="534" alt="Screen Shot 2023-07-24 at 2 23 00 PM" src="https://github.com/backdrop/backdrop-issues/assets/9938978/a64b9a37-13d2-418c-a2de-8a83ad0c1b87">
  

## Proposed solution

Be able to manually add the Dashboard context to layouts with paths other than `admin/dashboard`. This would allow other layouts to access dashboard-specific blocks.

## Alternatives that have been considered

None

## Additional information

According to a [Zulip comment](https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/Dashboard/near/378252421) by @docwilmot:
> In [layout.context.admin](https://github.com/backdrop/backdrop/blob/f32837d9a1ae3dcbc3bb0e27274bbd051ed89a95/core/modules/layout/layout.context.admin.inc#L81) there is this line `if (!in_array($key, array('string', 'overrides_path', 'dashboard')))` which prevents this ...There isn't any real reason not to be able to add dashboard contexts. We need an issue to remove that exception.

## Draft of feature description for Press Release (1 paragraph at most)

Backdrop now includes the ability to add the Dashboard context to layouts with any paths.

**Comment by docwilmot** [comment]
PR https://github.com/backdrop/backdrop/pull/4496 allows custom layouts to have dashboard context, which fulfills the actual issue requirement, but there are a couple specific conditions that apply to the default Dashboard path (`admin/dashboard`) that we could perhaps discuss how/if to apply to custom dashboards. 
- Should we  add a settings page to individual custom Dashboards?
- How to determine permissions for Dashboards?
- Should we implement `hook_admin_paths()`?

Maybe other questions.

**Comment by docwilmot** [comment]
I forget again how things get into a milestone. Help please?

**Comment by argiepiano** [comment]
I believe once the PR is reviewed and tested, the reviewer will add the actual milestone (in this case, probably 1.26.1 since it's a task), and remove the "candidate" tag.

I see the PR has a few coding standard failures. Can you fix that? I'll try to test tonight.

**Comment by bugfolder** [comment]
> I forget again how things get into a milestone. Help please?

More details at https://docs.backdropcms.org/documentation/adding-milestones-to-issues.

**Comment by argiepiano** [comment]
@docwilmot, I've done some testing - very nice! I encountered one problem: The Dashboard context cannot be removed once you add it. After clicking remove, the throbber spins for a couple seconds, but nothing gets removed. The db log shows `TypeError: Cannot access offset of type string on string in layout_settings_form_context_remove() (line 384 of /Users/XXX/Sites/localhost/XXX/core/modules/layout/layout.context.admin.inc).`

**Comment by docwilmot** [comment]
Thanks, on my local it removes properly, but I missed the watchdog error. Should be fixed now.

**Comment by argiepiano** [comment]
Thanks for fixing the issue. I've tested and it works great!

I do have a question about a line of code that determines if the Configure link should be shown:

`if ($layout_context->plugin != 'dashboard') {}`

I think hard-coding this condition works OK in this case, since this is the only plugin that doesn't require that. configure link. However I'm wondering if it may be a good idea to add one more key to the plugin definition hook (`hook_layout_context_info()`) that will indicate if this is a configurable context.

**Comment by argiepiano** [comment]
I've created a new issue to avoid hardcoding that `if`. Issue: https://github.com/backdrop/backdrop-issues/issues/6238

There is a different way to avoid hardcoding this check without adding a key - basically check if the context class  for the context plugin overrides `LayoutContext::form()` (which is an empty method in that class). If it doesn't we can safely assume that we don't need the `Configure` link.

Check the code sample I put in that issue and let me know what you think.

**Comment by docwilmot** [comment]
@argiepiano this sounds so familiar, I'm pretty sure I considered it but opted to leave for a separate issue. Should we finish this and look into configurable contexts separately?

**Comment by argiepiano** [comment]
> Should we finish this and look into configurable contexts separately?

Yes, I think that's a good idea. I marked this as RTBC and added a milestone.

**Comment by docwilmot** [comment]
Actually I forgot my own comment https://github.com/backdrop/backdrop-issues/issues/6182#issuecomment-1675969468. This PR allows dashboard contexts, but can we actually create custom dashboards? Do we need to decide on that before we allow custom contexts?

**Comment by argiepiano** [comment]
With this new context you can create a "faux dashboard" I guess, containing all those blocks, and then use a contrib module like LoginToboggan or Login Destination, to redirect users to that page upon login - so, it will act like a real dashboard. Anything else I missing? What else makes a "faux dashboard" different from the real dashboard other than the redirect setting at admin/dashboard/settings and the permission to access the dashboard?

**Comment by docwilmot** [comment]
> With this new context you can create a "faux dashboard" I guess, containing any of those blocks, and then use a contrib module like LoginToboggan or Login Destination, to redirect users to that page upon login - so, it will act like a real dashboard. Anything else I missing? What else makes a "faux dashboard" different from the real dashboard other than the redirect setting at admin/dashboard/settings and the permission to access the dashboard?
> 
> And regarding permissions to access - that can be added in the visibility conditions of the faux dashboard layout, by role perhaps.

Seems reasonable. Just wanted to make sure we ask the questions before users start expecting that dashboard contexts offer more than just the blocks. In either case the issue is satisfied by the PR, so RTBC it is! Thanks.

**Comment by quicksketch** [comment]
The PR looks good, I left one comment on it.

One issue I found is if you edit the existing `dashboard` layout and modify its path to something else (like `admin/dashboard2`) it throws a few errors. This does not appear to happen before this PR.

![image](https://github.com/backdrop/backdrop-issues/assets/100206/bb98f94e-f6d8-40a0-9a31-1b02927d0195)

I hesitate a little on this entire issue, since locking the Dashboard context to `admin/dashboard` was intentional. What kind of repercussions might this introduce?

**Comment by docwilmot** [comment]
Thanks for the review and the bug. Will check why this happens
I did also have some hesitation about the PR as it is now, as mentioned in https://github.com/backdrop/backdrop-issues/issues/6182#issuecomment-1675969468. I don't think that we should lock dashboard blocks to the one path as principle, but I agree we should further explore the implications to ensure that any new dashboard path works well.

**Comment by stpaultim** [comment]
Looking for clarification here. It looks like @quicksketch is suggesting that the code is almost there, but it sounds as if @docwilmot is looking to possibly a larger refactoring of this PR.

Is this close?

Also, this is listed on the 1.26.3 milestone but has the "milestone candidate - minor" tag. Is this a bugfix or minor candidate?
